File: skills/docforge/_shared/runtime/catalog/python/detect_profiles.py
#!/usr/bin/env python3
"""Detect Docforge shape, platform, framework, and concern profile candidates."""

from __future__ import annotations

import argparse
import fnmatch
import json
import re
import sys
from datetime import datetime, timezone
from pathlib import Path, PurePosixPath

from . import discovery_gate
from runtime.common.python import manifest_deps
from . import query_catalog

SKILL_ROOT = Path(__file__).resolve().parent.parent.parent.parent
DIMENSIONS = ["shapes", "platforms", "frameworks", "concerns"]
IGNORED = {
    ".git", ".codegraph", ".gitnexus", ".docforge", "node_modules",
    ".build", "build", "dist", "DerivedData", ".venv", "venv", "__pycache__",
}
MAX_FILES = 25000
MAX_CONTENT_BYTES = 1024 * 1024
MAX_TOTAL_CONTENT_BYTES = 32 * 1024 * 1024
MAX_EVIDENCE = 20
MAX_EXCERPTS = 8
MAX_EXCERPT_CHARS = 400
TEXT_SUFFIXES = {
    ".c", ".cc", ".cpp", ".cs", ".dart", ".go", ".gradle", ".h", ".hpp",
    ".csproj", ".html", ".java", ".js", ".json", ".jsx", ".kt", ".kts", ".m",
    ".md", ".mm", ".pbxproj", ".php", ".plist", ".properties", ".py", ".rb", ".rs", ".sh",
    ".sol", ".swift", ".toml", ".ts", ".tsx", ".txt", ".xml", ".yaml", ".yml",
}
TEXT_NAMES = {
    "CMakeLists.txt", "Dockerfile", "Gemfile", "Makefile", "Podfile",
    "requirements.txt",
}


def glob_regex(pattern: str) -> re.Pattern[str]:
    expression = []
    index = 0
    while index < len(pattern):
        char = pattern[index]
        if char == "*" and index + 1 < len(pattern) and pattern[index + 1] == "*":
            expression.append(".*")
            index += 2
        elif char == "*":
            expression.append("[^/]*")
            index += 1
        elif char == "?":
            expression.append("[^/]")
            index += 1
        else:
            expression.append(re.escape(char))
            index += 1
    return re.compile(f"^{''.join(expression)}$")


def matches_path(relative: str, pattern: str) -> bool:
    patterns = [pattern]
    if pattern.startswith("**/"):
        patterns.append(pattern[3:])
    basename = PurePosixPath(relative).name
    for candidate in patterns:
        regex = glob_regex(candidate)
        if regex.match(relative):
            return True
        if "/" not in candidate and regex.match(basename):
            return True
    return False


def signal_strength(signal: dict) -> str:
    explicit = signal.get("strength")
    if explicit in {"strong", "weak"}:
        return explicit
    kind = signal.get("kind")
    if kind == "dependency":
        return "strong"
    if kind == "content":
        return "weak"
    return "strong"


def cue_for_signal(signal: dict, relative: str = "") -> str:
    kind = signal.get("kind")
    if kind == "dependency":
        return f"dep:{signal.get('ecosystem', '')}:{signal.get('name', '')}"
    if kind == "content":
        token = (signal.get("contains") or "content").strip().lower().replace(" ", "-")
        return f"content:{token[:48]}"
    pattern = signal.get("pattern") or ""
    parts = [part for part in pattern.replace("**/", "").split("/") if part and "*" not in part]
    if parts:
        return f"path:{parts[-1].lower()}"
    if relative:
        fragments = [part for part in PurePosixPath(relative).parts if part not in {".", ".."}]
        if fragments:
            return f"path:{fragments[-2].lower() if len(fragments) > 1 else fragments[0].lower()}"
    return "path:unknown"


def inventory(repo: Path) -> list[tuple[str, Path]]:
    """Collect files under `repo` for profile-signal matching. Stops short of
    a nested repository's own contents — a subdirectory that is itself a
    separate git repository (its own .git dir or file, the same marker a
    submodule worktree uses) is a distinct project; its source must not be
    blended into this repo's own profile evidence."""
    found: list[tuple[str, Path]] = []

    def walk(directory: Path) -> None:
        if len(found) >= MAX_FILES:
            return
        for path in sorted(directory.iterdir(), key=lambda item: item.name):
            if path.name in IGNORED or path.is_symlink():
                continue
            if path.is_dir():
                git_marker = path / ".git"
                if git_marker.is_dir() or git_marker.is_file():
                    continue  # nested repo boundary; its evidence is its own
                walk(path)
            elif path.is_file():
                found.append((path.relative_to(repo).as_posix(), path))
            if len(found) >= MAX_FILES:
                return

    walk(repo)
    return found


def detect(
    repo: Path,
    *,
    persist: bool = True,
    files: list[tuple[str, Path]] | None = None,
) -> list[dict]:
    """Pass `files` to reuse an `inventory(repo)` the caller already has —
    the walk is the expensive part, and several callers need both."""
    profiles = query_catalog.load_profiles()
    if files is None:
        files = inventory(repo)
    dependencies = manifest_deps.extract_dependencies(files)
    if persist:
        persist_manifest_deps(repo, dependencies)
    cache: dict[Path, str] = {}
    cached_bytes = 0
    results: list[dict] = []
    for dimension in DIMENSIONS:
        for profile in profiles[dimension]:
            evidence: list[str] = []
            matched_strengths: set[str] = set()
            cues: list[str] = []
            for signal in profile.get("signals", []):
                strength = signal_strength(signal)
                if signal["kind"] == "dependency":
                    ecosystem = signal.get("ecosystem", "")
                    key = manifest_deps.normalize(ecosystem, signal.get("name", ""))
                    for manifest_path in dependencies.get(ecosystem, {}).get(key, []):
                        evidence.append(manifest_path)
                        matched_strengths.add(strength)
                        cues.append(cue_for_signal(signal))
                    continue
                for relative, path in files:
                    if not matches_path(relative, signal["pattern"]):
                        continue
                    if signal["kind"] == "content":
                        size = path.stat().st_size
                        if (
                            path.suffix.lower() not in TEXT_SUFFIXES
                            and path.name not in TEXT_NAMES
                        ):
                            continue
                        if size > MAX_CONTENT_BYTES:
                            continue
                        if path not in cache:
                            if cached_bytes + size > MAX_TOTAL_CONTENT_BYTES:
                                continue
                            cache[path] = path.read_text(encoding="utf-8", errors="replace")
                            cached_bytes += size
                        if signal.get("contains", "") not in cache[path]:
                            continue
                    evidence.append(relative)
                    matched_strengths.add(strength)
                    cues.append(cue_for_signal(signal, relative))
            evidence = sorted(set(evidence))
            if not evidence:
                continue
            cue_list = sorted(set(cues))
            has_strong = "strong" in matched_strengths
            match_strength = "strong" if has_strong else "weak"
            confidence = "confirmed" if has_strong else "candidate"
            results.append({
                "dimension": dimension,
                "id": profile["id"],
                "confidence": confidence,
                "evidence": evidence[:MAX_EVIDENCE],
                "match_strength": match_strength,
                "cues": cue_list,
                "ambiguous_with": [],
            })
    _attach_ambiguous_with(results)
    return results


def _attach_ambiguous_with(results: list[dict]) -> None:
    by_cue: dict[str, list[dict]] = {}
    for item in results:
        for cue in item.get("cues", []):
            if cue.startswith("path:") or cue.startswith("content:"):
                by_cue.setdefault(cue, []).append(item)
    for item in results:
        peers = []
        seen = set()
        for cue in item.get("cues", []):
            for peer in by_cue.get(cue, []):
                key = (peer["dimension"], peer["id"])
                if key == (item["dimension"], item["id"]) or key in seen:
                    continue
                if peer.get("match_strength") == "strong" and item.get("match_strength") == "strong":
                    continue
                seen.add(key)
                peers.append({
                    "dimension": peer["dimension"],
                    "id": peer["id"],
                    "confidence": peer["confidence"],
                    "cue": cue,
                })
        peers.sort(key=lambda row: (row["dimension"], row["id"], row.get("cue", "")))
        item["ambiguous_with"] = peers


def _dependency_summary(dependencies: dict) -> dict:
    summary = {}
    for ecosystem in sorted(dependencies):
        summary[ecosystem] = sorted(dependencies[ecosystem].keys())
    return summary


def _excerpts(repo: Path, evidence_paths: list[str], files: list[tuple[str, Path]]) -> list[dict]:
    index = {relative: path for relative, path in files}
    excerpts = []
    for relative in evidence_paths:
        if len(excerpts) >= MAX_EXCERPTS:
            break
        path = index.get(relative)
        if path is None:
            continue
        if path.suffix.lower() not in TEXT_SUFFIXES and path.name not in TEXT_NAMES:
            continue
        try:
            size = path.stat().st_size
        except OSError:
            continue
        if size > MAX_CONTENT_BYTES:
            continue
        text = path.read_text(encoding="utf-8", errors="replace")[:MAX_EXCERPT_CHARS]
        excerpts.append({"path": relative, "text": text, "max_chars": MAX_EXCERPT_CHARS})
    return excerpts


def emit_gate_pack(repo: Path) -> dict:
    profiles = query_catalog.load_profiles()
    files = inventory(repo)
    dependencies = manifest_deps.extract_dependencies(files)
    detections = detect(repo, files=files)
    strong = [item for item in detections if item["confidence"] == "confirmed"]
    weak = [item for item in detections if item["confidence"] == "candidate"]
    cue_map: dict[str, dict] = {}
    for item in detections:
        for cue in item.get("cues", []):
            bucket = cue_map.setdefault(cue, {
                "cue": cue,
                "surface": item["evidence"][0] if item["evidence"] else cue,
                "kind": "dependency" if cue.startswith("dep:") else (
                    "content_keyword" if cue.startswith("content:") else "path_fragment"
                ),
                "candidate_profiles": [],
            })
            entry = {
                "dimension": item["dimension"],
                "id": item["id"],
                "why": (
                    "strong signal"
                    if item.get("match_strength") == "strong"
                    else "weak path or content signal"
                ),
            }
            if entry not in bucket["candidate_profiles"]:
                bucket["candidate_profiles"].append(entry)
    # Offer persistence / ai-ml as peers for shared path nouns when weak cues fire.
    peer_concerns = ["persistence", "ai-ml"]
    for cue, bucket in cue_map.items():
        if not cue.startswith("path:"):
            continue
        present = {(row["dimension"], row["id"]) for row in bucket["candidate_profiles"]}
        for concern_id in peer_concerns:
            key = ("concerns", concern_id)
            if key in present:
                continue
            bucket["candidate_profiles"].append({
                "dimension": "concerns",
                "id": concern_id,
                "why": "catalog concern available; unconfirmed",
            })
    cues = sorted(cue_map.values(), key=lambda row: row["cue"])
    for bucket in cues:
        bucket["candidate_profiles"].sort(key=lambda row: (row["dimension"], row["id"]))
    evidence_paths = []
    for item in detections:
        for path in item["evidence"]:
            if path not in evidence_paths:
                evidence_paths.append(path)
    catalog_ids = {
        dimension: sorted(profile["id"] for profile in profiles[dimension])
        for dimension in [*DIMENSIONS, "audiences"]
    }
    query_hints = {}
    for dimension in DIMENSIONS:
        for profile in profiles[dimension]:
            if profile.get("query_hints"):
                query_hints[profile["id"]] = list(profile["query_hints"])
    # Side effect for WRITE: tech-stack reads this scratch instead of re-deriving.
    persist_manifest_deps(repo, dependencies)
    # Lazy import: common.scale imports this module, so a module-level import
    # here would be circular. The pack reuses the same walk and extraction.
    from runtime.common.python import scale as scale_module

    return {
        "repo": str(repo.resolve()),
        "detections": detections,
        "strong_detections": strong,
        "weak_detections": weak,
        "cues": cues,
        "excerpts": _excerpts(repo, evidence_paths, files),
        "dependencies": _dependency_summary(dependencies),
        "scale": scale_module.compute_scale(repo, files=files, detections=detections, dependencies=dependencies),
        "catalog_ids": catalog_ids,
        "query_hints": query_hints,
        "cue_hints": query_catalog.load_index().get("cue_hints", []),
        "needs_gate": discovery_gate.needs_gate(detections, cues),
    }


def persist_manifest_deps(repo: Path, dependencies: dict) -> Path:
    """Write manifest dependency rows for the tech-stack WRITE step."""
    scratch = repo / ".docforge" / "scratch"
    scratch.mkdir(parents=True, exist_ok=True)
    path = scratch / "manifest-deps.json"
    path.write_text(
        json.dumps(
            {
                "generated_at": datetime.now(timezone.utc)
                .replace(microsecond=0)
                .isoformat(),
                "dependencies": dependencies,
            },
            indent=2,
            ensure_ascii=False,
        )
        + "\n",
        encoding="utf-8",
    )
    return path


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--repo", required=True, type=Path)
    parser.add_argument("--json", action="store_true")
    parser.add_argument(
        "--emit-gate-pack",
        action="store_true",
        help="emit bounded discovery gate pack for agent interpretation",
    )
    args = parser.parse_args()
    if not args.repo.is_dir():
        print(f"error: not a directory: {args.repo}", file=sys.stderr)
        return 2
    if args.emit_gate_pack:
        print(json.dumps(emit_gate_pack(args.repo), indent=2, ensure_ascii=False))
        return 0
    results = detect(args.repo)
    if args.json:
        print(json.dumps({
            "repo": str(args.repo.resolve()),
            "detections": results,
        }, indent=2, ensure_ascii=False))
        return 0
    print(f"Profile detection for {args.repo.resolve()}")
    if not results:
        print("No profiles detected.")
        return 0
    for result in results:
        evidence = ", ".join(result["evidence"])
        print(
            f"{result['confidence'].upper():<9} "
            f"{result['dimension']:<10} {result['id']} "
            f"[{result['match_strength']}] — {evidence}"
        )
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: skills/docforge/_shared/runtime/catalog/js/detect_profiles.js
#!/usr/bin/env node
"use strict";
/** Detect Docforge shape, platform, framework, and concern profile candidates. */

const fs = require("fs");
const path = require("path");
const manifestDeps = require("../../common/js/manifest_deps.js");
const discoveryGate = require("./discovery_gate.js");
const queryCatalog = require("./query_catalog.js");

const SKILL_ROOT = path.resolve(fs.realpathSync(__dirname), "..", "..", "..");
const DIMENSIONS = ["shapes", "platforms", "frameworks", "concerns"];
const IGNORED = new Set([
  ".git", ".codegraph", ".gitnexus", ".docforge", "node_modules",
  ".build", "build", "dist", "DerivedData", ".venv", "venv", "__pycache__",
]);
const MAX_FILES = 25000;
const MAX_CONTENT_BYTES = 1024 * 1024;
const MAX_TOTAL_CONTENT_BYTES = 32 * 1024 * 1024;
const MAX_EVIDENCE = 20;
const MAX_EXCERPTS = 8;
const MAX_EXCERPT_CHARS = 400;
const TEXT_SUFFIXES = new Set([
  ".c", ".cc", ".cpp", ".cs", ".dart", ".go", ".gradle", ".h", ".hpp",
  ".csproj", ".html", ".java", ".js", ".json", ".jsx", ".kt", ".kts", ".m",
  ".md", ".mm", ".pbxproj", ".php", ".plist", ".properties", ".py", ".rb", ".rs", ".sh",
  ".sol", ".swift", ".toml", ".ts", ".tsx", ".txt", ".xml", ".yaml", ".yml",
]);
const TEXT_NAMES = new Set(["CMakeLists.txt", "Dockerfile", "Gemfile", "Makefile", "Podfile", "requirements.txt"]);

function compareText(left, right) {
  return Buffer.compare(Buffer.from(left, "utf8"), Buffer.from(right, "utf8"));
}
function globRegex(pattern) {
  let expression = "";
  for (let index = 0; index < pattern.length; index++) {
    const char = pattern[index];
    if (char === "*" && pattern[index + 1] === "*") {
      expression += ".*";
      index++;
    } else if (char === "*") expression += "[^/]*";
    else if (char === "?") expression += "[^/]";
    else expression += char.replace(/[\\^$+?.()|[\]{}]/g, "\\$&");
  }
  return new RegExp(`^${expression}$`);
}
function matchesPath(relative, pattern) {
  const patterns = pattern.startsWith("**/") ? [pattern, pattern.slice(3)] : [pattern];
  return patterns.some((candidate) => {
    const regex = globRegex(candidate);
    return regex.test(relative) || (!candidate.includes("/") && regex.test(path.posix.basename(relative)));
  });
}
function signalStrength(signal) {
  if (signal.strength === "strong" || signal.strength === "weak") return signal.strength;
  if (signal.kind === "dependency") return "strong";
  if (signal.kind === "content") return "weak";
  return "strong";
}
function cueForSignal(signal, relative) {
  if (signal.kind === "dependency") return `dep:${signal.ecosystem || ""}:${signal.name || ""}`;
  if (signal.kind === "content") {
    const token = String(signal.contains || "content").trim().toLowerCase().replace(/ /g, "-");
    return `content:${token.slice(0, 48)}`;
  }
  const pattern = signal.pattern || "";
  const parts = pattern.replace(/\*\*\//g, "").split("/").filter((part) => part && !part.includes("*"));
  if (parts.length) return `path:${parts[parts.length - 1].toLowerCase()}`;
  if (relative) {
    const fragments = relative.split("/").filter((part) => part && part !== "." && part !== "..");
    if (fragments.length) {
      return `path:${(fragments.length > 1 ? fragments[fragments.length - 2] : fragments[0]).toLowerCase()}`;
    }
  }
  return "path:unknown";
}
// Collect files under `repo` for profile-signal matching. Stops short of a
// nested repository's own contents — a subdirectory that is itself a
// separate git repository (its own .git dir or file, the same marker a
// submodule worktree uses) is a distinct project; its source must not be
// blended into this repo's own profile evidence.
function inventory(repo) {
  const found = [];
  function walk(directory) {
    if (found.length >= MAX_FILES) return;
    for (const entry of fs.readdirSync(directory, { withFileTypes: true }).sort((a, b) => compareText(a.name, b.name))) {
      if (IGNORED.has(entry.name)) continue;
      const target = path.join(directory, entry.name);
      if (entry.isDirectory()) {
        const gitMarker = path.join(target, ".git");
        if (fs.existsSync(gitMarker)) continue; // nested repo boundary; its evidence is its own
        walk(target);
      }
      else if (entry.isFile()) found.push([path.relative(repo, target).split(path.sep).join("/"), target]);
      if (found.length >= MAX_FILES) return;
    }
  }
  walk(repo);
  return found;
}
function attachAmbiguousWith(results) {
  const byCue = new Map();
  for (const item of results) {
    for (const cue of item.cues || []) {
      if (cue.startsWith("path:") || cue.startsWith("content:")) {
        if (!byCue.has(cue)) byCue.set(cue, []);
        byCue.get(cue).push(item);
      }
    }
  }
  for (const item of results) {
    const peers = [];
    const seen = new Set();
    for (const cue of item.cues || []) {
      for (const peer of byCue.get(cue) || []) {
        const key = `${peer.dimension}\0${peer.id}`;
        if (key === `${item.dimension}\0${item.id}` || seen.has(key)) continue;
        if (peer.match_strength === "strong" && item.match_strength === "strong") continue;
        seen.add(key);
        peers.push({
          dimension: peer.dimension,
          id: peer.id,
          confidence: peer.confidence,
          cue,
        });
      }
    }
    peers.sort((a, b) => compareText(`${a.dimension}:${a.id}:${a.cue}`, `${b.dimension}:${b.id}:${b.cue}`));
    item.ambiguous_with = peers;
  }
}
function persistManifestDeps(repo, dependencies) {
  const scratch = path.join(repo, ".docforge", "scratch");
  fs.mkdirSync(scratch, { recursive: true });
  const target = path.join(scratch, "manifest-deps.json");
  fs.writeFileSync(
    target,
    `${JSON.stringify(
      {
        generated_at: new Date().toISOString().replace(/\.\d{3}Z$/, "+00:00"),
        dependencies,
      },
      null,
      2,
    )}\n`,
  );
  return target;
}

// Pass `files` to reuse an `inventory(repo)` the caller already has — the walk
// is the expensive part, and several callers need both.
function detect(repo, persist = true, files = null) {
  const profiles = queryCatalog.loadProfiles();
  if (files === null) files = inventory(repo);
  const dependencies = manifestDeps.extractDependencies(files);
  if (persist) persistManifestDeps(repo, dependencies);
  const cache = new Map();
  let cachedBytes = 0;
  const results = [];
  for (const dimension of DIMENSIONS) {
    for (const profile of profiles[dimension]) {
      const evidence = new Set();
      const matchedStrengths = new Set();
      const cues = new Set();
      for (const signal of profile.signals || []) {
        const strength = signalStrength(signal);
        if (signal.kind === "dependency") {
          const ecosystem = signal.ecosystem || "";
          const key = manifestDeps.normalize(ecosystem, signal.name || "");
          for (const manifestPath of (dependencies[ecosystem] || {})[key] || []) {
            evidence.add(manifestPath);
            matchedStrengths.add(strength);
            cues.add(cueForSignal(signal));
          }
          continue;
        }
        for (const [relative, target] of files) {
          if (!matchesPath(relative, signal.pattern)) continue;
          if (signal.kind === "content") {
            const size = fs.statSync(target).size;
            if (!TEXT_SUFFIXES.has(path.extname(target).toLowerCase()) && !TEXT_NAMES.has(path.basename(target))) continue;
            if (size > MAX_CONTENT_BYTES) continue;
            if (!cache.has(target)) {
              if (cachedBytes + size > MAX_TOTAL_CONTENT_BYTES) continue;
              cache.set(target, fs.readFileSync(target, "utf8"));
              cachedBytes += size;
            }
            if (!cache.get(target).includes(signal.contains || "")) continue;
          }
          evidence.add(relative);
          matchedStrengths.add(strength);
          cues.add(cueForSignal(signal, relative));
        }
      }
      if (!evidence.size) continue;
      const paths = [...evidence].sort(compareText);
      const cueList = [...cues].sort(compareText);
      const hasStrong = matchedStrengths.has("strong");
      results.push({
        dimension,
        id: profile.id,
        confidence: hasStrong ? "confirmed" : "candidate",
        evidence: paths.slice(0, MAX_EVIDENCE),
        match_strength: hasStrong ? "strong" : "weak",
        cues: cueList,
        ambiguous_with: [],
      });
    }
  }
  attachAmbiguousWith(results);
  return results;
}
function dependencySummary(dependencies) {
  const summary = {};
  for (const ecosystem of Object.keys(dependencies).sort(compareText)) {
    summary[ecosystem] = Object.keys(dependencies[ecosystem]).sort(compareText);
  }
  return summary;
}
function excerpts(repo, evidencePaths, files) {
  const index = new Map(files);
  const out = [];
  for (const relative of evidencePaths) {
    if (out.length >= MAX_EXCERPTS) break;
    const target = index.get(relative);
    if (!target) continue;
    if (!TEXT_SUFFIXES.has(path.extname(target).toLowerCase()) && !TEXT_NAMES.has(path.basename(target))) continue;
    let size;
    try { size = fs.statSync(target).size; } catch { continue; }
    if (size > MAX_CONTENT_BYTES) continue;
    const text = fs.readFileSync(target, "utf8").slice(0, MAX_EXCERPT_CHARS);
    out.push({ path: relative, text, max_chars: MAX_EXCERPT_CHARS });
  }
  return out;
}
function emitGatePack(repo) {
  const profiles = queryCatalog.loadProfiles();
  const files = inventory(repo);
  const dependencies = manifestDeps.extractDependencies(files);
  const detections = detect(repo, true, files);
  const strong = detections.filter((item) => item.confidence === "confirmed");
  const weak = detections.filter((item) => item.confidence === "candidate");
  const cueMap = new Map();
  for (const item of detections) {
    for (const cue of item.cues || []) {
      if (!cueMap.has(cue)) {
        cueMap.set(cue, {
          cue,
          surface: item.evidence[0] || cue,
          kind: cue.startsWith("dep:") ? "dependency" : cue.startsWith("content:") ? "content_keyword" : "path_fragment",
          candidate_profiles: [],
        });
      }
      const bucket = cueMap.get(cue);
      const entry = {
        dimension: item.dimension,
        id: item.id,
        why: item.match_strength === "strong" ? "strong signal" : "weak path or content signal",
      };
      if (!bucket.candidate_profiles.some((row) => row.dimension === entry.dimension && row.id === entry.id)) {
        bucket.candidate_profiles.push(entry);
      }
    }
  }
  const peerConcerns = ["persistence", "ai-ml"];
  for (const [, bucket] of cueMap) {
    if (!bucket.cue.startsWith("path:")) continue;
    const present = new Set(bucket.candidate_profiles.map((row) => `${row.dimension}\0${row.id}`));
    for (const concernId of peerConcerns) {
      const key = `concerns\0${concernId}`;
      if (present.has(key)) continue;
      bucket.candidate_profiles.push({
        dimension: "concerns",
        id: concernId,
        why: "catalog concern available; unconfirmed",
      });
    }
  }
  const cues = [...cueMap.values()].sort((a, b) => compareText(a.cue, b.cue));
  for (const bucket of cues) {
    bucket.candidate_profiles.sort((a, b) => compareText(`${a.dimension}:${a.id}`, `${b.dimension}:${b.id}`));
  }
  const evidencePaths = [];
  for (const item of detections) {
    for (const evidencePath of item.evidence) {
      if (!evidencePaths.includes(evidencePath)) evidencePaths.push(evidencePath);
    }
  }
  const catalogIds = {};
  for (const dimension of [...DIMENSIONS, "audiences"]) {
    catalogIds[dimension] = profiles[dimension].map((profile) => profile.id).sort(compareText);
  }
  const queryHints = {};
  for (const dimension of DIMENSIONS) {
    for (const profile of profiles[dimension]) {
      if (profile.query_hints && profile.query_hints.length) queryHints[profile.id] = [...profile.query_hints];
    }
  }
  return {
    repo: fs.realpathSync(repo),
    detections,
    strong_detections: strong,
    weak_detections: weak,
    cues,
    excerpts: excerpts(repo, evidencePaths, files),
    dependencies: dependencySummary(dependencies),
    // Lazy require: common/scale requires this module, so a module-level
    // require here would be circular. The pack reuses the same walk and
    // extraction.
    scale: require("../../common/js/scale.js").computeScale(repo, files, detections, dependencies),
    catalog_ids: catalogIds,
    query_hints: queryHints,
    cue_hints: queryCatalog.loadIndex().cue_hints || [],
    needs_gate: discoveryGate.needsGate(detections, cues),
  };
}
function parseArgs(argv) {
  const result = { json: false, emitGatePack: false };
  for (let index = 0; index < argv.length; index++) {
    const token = argv[index];
    if (token === "--json") result.json = true;
    else if (token === "--emit-gate-pack") result.emitGatePack = true;
    else if (token === "--repo") {
      if (index + 1 >= argv.length || argv[index + 1].startsWith("--")) throw new Error("option requires a value: --repo");
      result.repo = argv[++index];
    } else if (token === "-h" || token === "--help") result.help = true;
    else throw new Error(`unknown option: ${token}`);
  }
  return result;
}
function usage() {
  console.log("usage: detect_profiles.js --repo <path> [--json] [--emit-gate-pack]");
}
function main() {
  let args;
  try {
    args = parseArgs(process.argv.slice(2));
  } catch (error) {
    usage();
    console.error(`error: ${error.message}`);
    return 2;
  }
  if (args.help) {
    usage();
    return 0;
  }
  if (!args.repo || !fs.existsSync(args.repo) || !fs.statSync(args.repo).isDirectory()) {
    console.error(`error: not a directory: ${args.repo || ""}`);
    return 2;
  }
  const repo = fs.realpathSync(args.repo);
  if (args.emitGatePack) {
    console.log(JSON.stringify(emitGatePack(repo), null, 2));
    return 0;
  }
  const results = detect(repo);
  if (args.json) {
    console.log(JSON.stringify({ repo, detections: results }, null, 2));
    return 0;
  }
  console.log(`Profile detection for ${repo}`);
  if (!results.length) {
    console.log("No profiles detected.");
    return 0;
  }
  for (const result of results) {
    console.log(`${result.confidence.toUpperCase().padEnd(9)} ${result.dimension.padEnd(10)} ${result.id} [${result.match_strength}] — ${result.evidence.join(", ")}`);
  }
  return 0;
}
if (require.main === module) process.exit(main());
module.exports = { detect, emitGatePack, inventory, main };
